openwisp_monitoring/monitoring/base/models.py

```diff
@@ -154,48 +156,97 @@ def _get_time(self, time):
         """
-        checks if the threshold is crossed
-        and notifies users accordingly
+        Sets the value of "is_healthy" field if "value"
+        crosses threshold defined in "alert_settings".
+        Returns "True" if "is_healthy" field is changed.
+        Otherwise, returns "None".
         """
-        try:
-            alert_settings = self.alertsettings
-        except ObjectDoesNotExist:
+        crossed = alert_settings._value_crossed(value)
+        if (not crossed and self.is_healthy) or (crossed and self.is_healthy is False):
             return
+        # problem: not within threshold limit
+        elif crossed and self.is_healthy in [True, None]:
+            self.is_healthy = False
+        # ok: returned within threshold limit
+        elif not crossed and self.is_healthy in [False, None]:
+            self.is_healthy = True
+        return True
+
+    def _set_is_tolerance_healthy(
+        self, alert_settings, value, time, retention_policy, send_alert
+    ):
+        """
+        Sets the value of "is_tolerance_healthy" if "value"
+        crosses threshold for tolerance defined in alert_settings.
+        It also sends the notification if required.
+        Returns "None" if value of "is_tolerance_healthy" is unchanged.
+        Returns "True" if it is the first metric write within threshold.
+        Returns "False" in other cases.
+        """
         time = self._get_time(time)
         crossed = alert_settings._is_crossed_by(value, time, retention_policy)
         first_time = False
         # situation has not changed
-        if (not crossed and self.is_healthy) or (crossed and self.is_healthy is False):
+        if (not crossed and self.is_tolerance_healthy) or (
+            crossed and self.is_tolerance_healthy is False
+        ):
             return
         # problem: not within threshold limit
-        elif crossed and self.is_healthy in [True, None]:
-            if self.is_healthy is None:
+        elif crossed and self.is_tolerance_healthy in [True, None]:
+            if self.is_tolerance_healthy is None:
                 first_time = True
-            self.is_healthy = False
+            self.is_tolerance_healthy = False
             notification_type = f'{self.configuration}_problem'
         # ok: returned within threshold limit
-        elif not crossed and self.is_healthy is False:
-            self.is_healthy = True
+        elif not crossed and self.is_tolerance_healthy is False:
+            self.is_tolerance_healthy = True
             notification_type = f'{self.configuration}_recovery'
         # First metric write within threshold
-        elif not crossed and self.is_healthy is None:
-            self.is_healthy = True
+        elif not crossed and self.is_tolerance_healthy is None:
+            self.is_tolerance_healthy = True
             first_time = True
-        self.save()
+
+        # Don't raise alert if it is the first metric within threshold.
```

Comment:
I would write something like:

```
If we got to this point, it means we have to send an alert, because the metric
has been crossed for more than the tolerated amount of time.
There's one exception though: if the device is new, its status will be unknown
and the metric will become healthy for the first time, in this case we do not need
to send an alert.
```